packages/shared/src/hooks/squads/useSquadActions.ts

```diff
@@ -0,0 +1,78 @@
+import { useMemo } from 'react';
+import {
+  useInfiniteQuery,
+  UseInfiniteQueryResult,
+  useMutation,
+  useQueryClient,
+} from 'react-query';
+import request from 'graphql-request';
+import {
+  SQUAD_MEMBERS_QUERY,
+  SquadEdgesData,
+  updateSquadMemberRole,
+  verifyPermission,
+} from '../../graphql/squads';
+import { graphqlUrl } from '../../lib/config';
+import { SourceMember, SourcePermissions, Squad } from '../../graphql/sources';
+
+export interface UseSquadActions {
+  onUpdateRole?: typeof updateSquadMemberRole;
+  membersQueryResult?: UseInfiniteQueryResult<SquadEdgesData>;
+  members?: SourceMember[];
+  verifyPermission: (permission: SourcePermissions) => boolean;
+}
+
+interface UseSquadActionsProps {
+  squad: Squad;
+  membersQueryEnabled?: boolean;
+}
+
+export const useSquadActions = ({
+  squad,
+  membersQueryEnabled,
+}: UseSquadActionsProps): UseSquadActions => {
+  const client = useQueryClient();
+  const { mutateAsync: onUpdateRole } = useMutation(updateSquadMemberRole, {
+    onSuccess: (_, { sourceId }) =>
+      client.invalidateQueries(['squadMembers', sourceId]),
+  });
+
+  const membersQueryKey = ['squadMembers', squad?.id];
+  const membersQueryResult = useInfiniteQuery<SquadEdgesData>(
+    membersQueryKey,
+    ({ pageParam }) =>
+      request(graphqlUrl, SQUAD_MEMBERS_QUERY, {
+        id: squad?.id,
+        after: pageParam,
+      }),
+    {
+      enabled: !!squad?.id && membersQueryEnabled,
+      getNextPageParam: (lastPage) =>
+        lastPage?.sourceMembers?.pageInfo?.hasNextPage &&
+        lastPage?.sourceMembers?.pageInfo?.endCursor,
+    },
+  );
+
+  const members = useMemo(
+    () =>
+      membersQueryResult.data?.pages
+        .map((page) =>
+          page.sourceMembers.edges.map(({ node }) => ({ ...node, page })),
+        )
+        .flat() ?? [],
+    [membersQueryResult],
+  );
+
+  const checkHasPermission = (permission: SourcePermissions) =>
+    verifyPermission(squad, permission);
+
+  return useMemo(
+    () => ({
+      onUpdateRole,
+      membersQueryResult,
+      members,
+      verifyPermission: checkHasPermission,
+    }),
+    [onUpdateRole, membersQueryResult, squad, membersQueryEnabled],
+  );
```

Comment:
Still some minor inconsistencies for variables inside dependency array:

Missing from array:
- members
- checkHasPermission

Can be removed from array:
- membersQueryEnabled
- squad

Reply:
> membersQueryEnabled

Oh, right. This can be removed. Thanks!

> members

The `members` have a dependency from `membersQueryResult` which is already included here, when that changes, both `members` and the function in this `useMemo` will get executed to return the updated value.

> checkHasPermission

This is not a stable dependency since this is a regular function unless we use `useCallback` which I think is unnecessary, that's why we need the `squad` as a dependency, so when the `squad` changes, this `useMemo` will fire the function and then get the newly re-created `checkHasPermission` function.

Reply:
> The members have a dependency from membersQueryResult which is already included here, when that dependency changes, both members and the function in this useMemo will get executed to return the updated value.

Hmm there is no point in not including `members` here then, since it will act the same as you said. 😃 

Imagine down the line members is no longer connected to membersQueryResult, what then? We would get bugs where members has stale values due to it not being included in deps array. If you want we can also talk in gather about this if you find it faster to close the discussion.

Reply:
Good point, that makes devex better, but the concern is, `members` will always be based on the members query result.

Probably, what we can do is remove the variable declaration completely since this is not used anywhere inside this hook and just make it in-line with the return of `useMemo`. That way we can reduce the places we check. WDYT?

Reply:
@sshanzel yep, this is the way 👌